File: current_session/c++/122.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int maxProfit(vector<int>& prices) {
    // use dp
    // buy and sell
    long buy= long(-prices[0]), sell = 0;
    for (int i=1; i<prices.size(); ++i) {
      int newBuy = max(buy, sell-prices[i]);
      int newSell = max(sell, buy+prices[i]);
      buy = newBuy;
      sell = newSell;
    }
    return sell;

  }
};

File: current_session/python/122.py
from typing import List
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy, notBuy = -1*prices[0], 0
        for i in range(1, len(prices)):
            buy, notBuy = max(buy, notBuy-prices[i]), max(notBuy, buy+prices[i])
        
        return notBuy